In [1]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

### Question 1

What's the version of NumPy that you installed? 

In [2]:
np.__version__

'1.18.5'

### Question 2

What's the version of Pandas? 

In [3]:
pd.__version__

'1.3.0'

In [4]:
!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

--2021-09-07 19:00:01--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
raw.githubusercontent.com (raw.githubusercontent.com) çözümleniyor... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443 bağlanılıyor... bağlantı kuruldu.
HTTP isteği gönderildi, cevap bekleniyor... 200 OK
Uzunluk: 1475504 (1,4M) [text/plain]
Kayıt yeri: `data.csv'

data.csv            100%[===================>]   1,41M  2,53MB/s    süre 0,6s  

2021-09-07 19:00:03 (2,53 MB/s) - `data.csv' kaydedildi [1475504/1475504]



In [5]:
!ls

data.csv  MLZoomCamp-Homework-1.ipynb


In [6]:
data = pd.read_csv("data.csv")
data.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

### Question 3

What's the average price of BMW cars in the dataset?

In [32]:
data["Make"].unique()

array(['BMW', 'Audi', 'FIAT', 'Mercedes-Benz', 'Chrysler', 'Nissan',
       'Volvo', 'Mazda', 'Mitsubishi', 'Ferrari', 'Alfa Romeo', 'Toyota',
       'McLaren', 'Maybach', 'Pontiac', 'Porsche', 'Saab', 'GMC',
       'Hyundai', 'Plymouth', 'Honda', 'Oldsmobile', 'Suzuki', 'Ford',
       'Cadillac', 'Kia', 'Bentley', 'Chevrolet', 'Dodge', 'Lamborghini',
       'Lincoln', 'Subaru', 'Volkswagen', 'Spyker', 'Buick', 'Acura',
       'Rolls-Royce', 'Maserati', 'Lexus', 'Aston Martin', 'Land Rover',
       'Lotus', 'Infiniti', 'Scion', 'Genesis', 'HUMMER', 'Tesla',
       'Bugatti'], dtype=object)

In [33]:
BMW = data[data["Make"] == "BMW"]
mean = np.mean(BMW["MSRP"])
print("Average prices of BMW cars: ", mean)

Average prices of BMW cars:  61546.76347305389


### Question 4

Select a subset of cars after year 2015. How many of them have missing values for Engine HP?

In [34]:
after_2015 = data[data["Year"] > 2015]
after_2015

Make       Model  Year                Engine Fuel Type  Engine HP  \
32     FIAT  124 Spider  2017  premium unleaded (recommended)      160.0   
33     FIAT  124 Spider  2017  premium unleaded (recommended)      160.0   
34     FIAT  124 Spider  2017  premium unleaded (recommended)      160.0   
41      BMW    2 Series  2016     premium unleaded (required)      240.0   
42      BMW    2 Series  2016     premium unleaded (required)      240.0   
...     ...         ...   ...                             ...        ...   
11879   GMC       Yukon  2017        flex-fuel (unleaded/E85)      355.0   
11880   GMC       Yukon  2017        flex-fuel (unleaded/E85)      355.0   
11900   BMW          Z4  2016     premium unleaded (required)      300.0   
11901   BMW          Z4  2016     premium unleaded (required)      240.0   
11902   BMW          Z4  2016     premium unleaded (required)      335.0   

       Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
32                  4.0            MANUAL  rear wheel drive              2.0   
33                  4.0            MANUAL  rear wheel drive              2.0   
34                  4.0            MANUAL  rear wheel drive              2.0   
41                  4.0         AUTOMATIC  rear wheel drive              2.0   
42                  4.0         AUTOMATIC  rear wheel drive              2.0   
...                 ...               ...               ...              ...   
11879               8.0         AUTOMATIC  four wheel drive              4.0   
11880               8.0         AUTOMATIC  rear wheel drive              4.0   
11900               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   
11901               4.0            MANUAL  rear wheel drive              2.0   
11902               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   

               Market Category Vehicle Size Vehicle Style  highway MPG  \
32                 Performance      Compact   Convertible           35   
33                 Performance      Compact   Convertible           35   
34                 Performance      Compact   Convertible           35   
41          Luxury,Performance      Compact         Coupe           35   
42                      Luxury      Compact   Convertible           34   
...                        ...          ...           ...          ...   
11879    Flex Fuel,Performance        Large       4dr SUV           22   
11880    Flex Fuel,Performance        Large       4dr SUV           23   
11900  Luxury,High-Performance      Compact   Convertible           24   
11901       Luxury,Performance      Compact   Convertible           34   
11902  Luxury,High-Performance      Compact   Convertible           24   

       city mpg  Popularity   MSRP  
32           26         819  27495  
33           26         819  24995  
34           26         819  28195  
41           23        3916  32850  
42           23        3916  38650  
...         ...         ...    ...  
11879        16         549  60315  
11880        16         549  48530  
11900        17        3916  57500  
11901        22        3916  49700  
11902        17        3916  66350  

[3825 rows x 16 columns]

In [35]:
after_2015["Engine HP"].isna().sum()

0

There are no missing values for Engine HP in cars for year 2015.

### Question 5

* Calculate the average "Engine HP" in the dataset. 

In [26]:
mean_enginehp = data["Engine HP"].mean()
mean_enginehp

249.38607007176023

* Use the `fillna` method and to fill the missing values in "Engine HP" with the mean value from the previous step.

In [25]:
data["Engine HP"].isna().sum()

69

In [27]:
data["Engine HP"].fillna(mean_enginehp, inplace = True)

In [29]:
data["Engine HP"].isna().sum()

0

* Now, calcualte the average of "Engine HP" again.

In [30]:
new_mean_enginehp = data["Engine HP"].mean()
new_mean_enginehp

249.38607007176023

* Has it changed?

In [31]:
print(round(mean_enginehp))
print(round(new_mean_enginehp))

249.0
249.0


The mean hasn't changed.

### Question 6

* Select all the "Rolls-Royce" cars from the dataset.

In [36]:
rolls_royce = data[data["Make"] == "Rolls-Royce"]

* Select only columns "Engine HP", "Engine Cylinders", "highway MPG".

In [37]:
new_df_rolls = rolls_royce[["Engine HP", "Engine Cylinders", "highway MPG"]]

In [39]:
new_df_rolls.head()

Engine HP  Engine Cylinders  highway MPG
2921      325.0               8.0           15
3505      563.0              12.0           19
5275      563.0              12.0           21
5276      563.0              12.0           21
5277      563.0              12.0           21

* Now drop all duplicated rows using `drop_duplicates` method (you should get a dataframe with 7 rows).

In [68]:
duplicates = new_df_rolls.drop_duplicates()
duplicates

Engine HP  Engine Cylinders  highway MPG
2921       325.0               8.0           15
3505       563.0              12.0           19
5275       563.0              12.0           21
5279       563.0              12.0           20
7443       322.0              12.0           15
7553       453.0              12.0           19
11448      624.0              12.0           21

* Get the underlying NumPy array. Let's call it `X`.

In [69]:
X = duplicates.to_numpy()

In [70]:
X

array([[325.,   8.,  15.],
       [563.,  12.,  19.],
       [563.,  12.,  21.],
       [563.,  12.,  20.],
       [322.,  12.,  15.],
       [453.,  12.,  19.],
       [624.,  12.,  21.]])

* Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.

In [71]:
trans = X.T
XTX = np.dot(trans, X)
XTX

array([[1.754801e+06, 3.965600e+04, 6.519600e+04],
       [3.965600e+04, 9.280000e+02, 1.500000e+03],
       [6.519600e+04, 1.500000e+03, 2.454000e+03]])

* Invert `XTX`.

In [72]:
invert = np.linalg.inv(XTX)
invert

array([[ 5.17815728e-05,  9.06587044e-04, -1.92984188e-03],
       [ 9.06587044e-04,  1.05723058e-01, -8.87084092e-02],
       [-1.92984188e-03, -8.87084092e-02,  1.05900809e-01]])

* What's the sum of all the elements of the result?

In [73]:
invert.sum()

0.032212320677486125

### Question 7

* Create an array `y` with values `[1000, 1100, 900, 1200, 1000, 850, 1300]`.

In [74]:
y = np.array([1000, 1100, 900, 1200, 1000, 850, 1300])
y

array([1000, 1100,  900, 1200, 1000,  850, 1300])

* Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.

In [80]:
dot = np.dot(invert, trans)
w = np.dot(dot, y)
w

array([ 0.19989598, 31.02612262, 31.65378877])

* What's the value of the first element of `w`?.

    » 0.19989598